File: built-in plugins dsp/mb-eq-graphic.cpp
/**
 * MB Graphic EQ
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 10-band graphic EQ
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_EQ_GRAPHIC_H
#define MB_EQ_GRAPHIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbEqGraphic : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-eq-graphic";
    static constexpr const char* PLUGIN_NAME    = "MB Graphic EQ";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float band1 = 0f;  // range [-12, 12]
    float band2 = 0f;  // range [-12, 12]
    float band3 = 0f;  // range [-12, 12]
    float band4 = 0f;  // range [-12, 12]
    float band5 = 0f;  // range [-12, 12]
    };

    MbEqGraphic() = default;
    ~MbEqGraphic() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.band1 = std::clamp(params.band1, -12f, 12f);
        params.band2 = std::clamp(params.band2, -12f, 12f);
        params.band3 = std::clamp(params.band3, -12f, 12f);
        params.band4 = std::clamp(params.band4, -12f, 12f);
        params.band5 = std::clamp(params.band5, -12f, 12f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Graphic EQ
        return input;
    }
};

#endif // MB_EQ_GRAPHIC_H

File: built-in plugins dsp/mb-eq-graphic.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbEqGraphicPlugin: PluginDefinition = { id: 'mb-eq-graphic', slug: 'mb-eq-graphic', name: 'MB Graphic EQ', category: 'effect', type: 'eq', version: '1.0.0', description: '10-band graphic EQ', author: 'Max Booster', parameters: [{ id: 'band1', name: '31Hz', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'band2', name: '62Hz', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'band3', name: '125Hz', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'band4', name: '250Hz', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'band5', name: '500Hz', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }], defaultPreset: { band1: 0, band2: 0, band3: 0, band4: 0, band5: 0 } };

export default MbEqGraphicPlugin;
